nixos/zfs: only enable trim if zfs is enabled

Mic92 opened the pull request:
Also don't fail the service if there are no pools yet.
This might happen on installation ISOs.

See also: https://github.com/NixOS/nixpkgs/pull/69672#issuecomment-538793932
cc @delroth

This has been not tested yet.


<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

Mic92 committed:
nixos/zfs: only enable trim if zfs is enabled

Also don't fail the service if there are no pools yet.
This might happen on installation ISOs.

Mic92 committed:
nixos/zfs: avoid script derivation for trim service

Since we only have a single pipe we can save the overhead of building a derivation
when creating the zfs trim service file when building the system.

delroth reviewed (approved):
Tested that zpool-trim.service is now not started on my non-ZFS systems. Thanks!

Note that this makes the logic for zpool-trim different than the one implemented for the other ZFS services in that file. For example, `enableZfs` is defined to be enabled if `inuse || enableAutoSnapshots || enableAutoScrub`, without taking zpool-trim into account. IMHO `enableAutoSnapshots` and `enableAutoScrub` should be removed from the `enableZfs` boolean, but I don't have a clear understanding of the consequences or how much it would break.

delroth commented on nixos/modules/tasks/filesystems/zfs.nix:
```diff
@@ -552,15 +552,13 @@ in
-        script = ''
-          zpool list -H -o name | xargs -n1 zpool trim
-        '';
+        ExecStart = "${pkgs.runtimeShell} -c 'zpool list -H -o name | xargs --no-run-if-empty -n1 zpool trim'";
```
Should be serviceConfig.ExecStart.

Mic92 committed:
nixos/zfs: simplify logic for scrub/autosnapshot service

This makes them consistent with the way zfs.trim is enabled
and allow to enable them by default in future.

Mic92 committed:
nixos/zfs: mention trim support in the release notes

Mic92 commented:
> Note that this makes the logic for zpool-trim different than the one implemented for the other ZFS services in that file. For example, `enableZfs` is defined to be enabled if `inuse || enableAutoSnapshots || enableAutoScrub`, without taking zpool-trim into account. IMHO `enableAutoSnapshots` and `enableAutoScrub` should be removed from the `enableZfs` boolean, but I don't have a clear understanding of the consequences or how much it would break.

The Autosnapshot/autoscrub services are now only enabled if `zfs` is somehow used in the system.
